### ISSUE TITLE
DietPi-Software | Docker logging to RAM/journald

### ISSUE THREAD

**Issue by SaturnusDJ** [MichaIng/DietPi#2388]
Correct me if I am wrong but:

I found that docker containers log to  _/mnt/dietpi_userdata/docker-data/containers/[Container ID]/[Container ID]-json.log._ This is on non tmpfs, thus SD card in many cases.

DietPi by default has journald as logger log to _/run/log_, which is tmpfs.

Docker can logs can be passed to journald.
Therefor on a stock DietPi the /etc/docker/daemon.json file should look like:
```
{
    "data-root": "/mnt/dietpi_userdata/docker-data",
    "log-driver": "journald"
}
```
I am suggesting to discuss here if it would be a good idea to have this as default with DietPi.

Reporting this as bug since a) I consider it as design flaw on SD-card systems (as many DietPi users will have) for the many writes that can occur on e.g. info level logging, which is default in some applications, and b) as it is not in line with the logging philosophy presented in _dietpi-software_. Welcoming other's opinions on this.

**Comment by MichaIng** [comment]
@SaturnusDJ 
Thanks for sharing your suggestion.

Yeah if it's about me, I would move really all logging to systemd (I did on my system, where possible). It is by default on RAM, but can be as well (with custom rotation as desired) written permanently to disk. It is well sortable and you have full overview of what happened around certain events, instead of having the need to search several logs and compare timestamps... Sadly it is not even possible for all system and software logs. APT, dpkg and other system logs are still forcefully to /var/log (without moving them manually/hackish), and many software titles do not allow logging to syslog daemon.

The only advantage of disk log I see is that they can be read out easily on any other ext4 supporting system with text editor, without special systemd reader.

To resolve `/var/log` to disk, we have DietPi-RAMlog. But again, if the software supports journald logging, is does not make much sense to let it log to `/var/log` and mount this to RAM manually. And of course it does not apply to log dirs outside of /var/log.

In case of docker, the main question is, if users mostly wish to have container logs persistent or not. If there is some need to have this persistent, then at least `/mnt/dietpi_userdata/...` on SDcard systems is the better solution, since this should on production system anyway be moved to an external HDD/SSD, where you potentially have better performance and which deal muuuch better with constant I/O than SDcards.
____________
**To sum up: I would go with your journald solution and add info to our online docs, how to (re)apply persistent disk logging, if desired.**
@Fourdee 
What do you think?

**Comment by MichaIng** [comment]
@SaturnusDJ 
Since we just improved disk I/O in case of another software title, I think we can implement it for docker as well.

But two questions:
- It is not possible to move only the log file location, e.g. to /var/log, since it is always placed within docker-data, right?
- Could you give an example about the log amount?

E.g. if the amount of logs is high (more than ~10 entries each hour or something), then I personally would prefer it to log to a separate file in /var/log (DietPi-RAMlog, if enabled) instead of filling `journalctl` too much. But everything is better then dietpi_userdata (disk), this is where I totally agree.

**Comment by SaturnusDJ** [comment]
@MichaIng 
I am familiar with basic Docker functionality for some weeks now, not an expert. What I found is https://docs.docker.com/v17.09/engine/admin/logging/overview/#use-environment-variables-or-labels-with-logging-drivers.

_None_ would result in no logs.
_Json-file_ is default. It logs to a json-file in the root of the container meta volume. You would expect to be able to log to custom filesystem location, but I could find a single hint on the whole internet. A soft link would be an option but then a program needs to intercept the json-file and create a soft link each time a new container is created. Not nice.
_Syslog_ seems to be not running on DietPi?
_Journalct_, I don't know it well. It just worked for me. If this has options to log to folder, then that would come closest.
The other options require external programs, which extremely likely would make 'raw' logging to /var/log possible.

Then there is the option of using custom 'plug-in' logging drivers, yet I couldn't find one.

**Comment by MichaIng** [comment]
@SaturnusDJ 
Thanks for info.

> Syslog seems to be not running on DietPi?

We follow the (modern) Debian defaults here and have journald as syslog daemon. However you can choose rsyslog + logrotate logging mode via dietpi-software, which will disable RAMlog (for /var/log) as well, otherwise rsyslog does not make much sense. Of course you could install rsyslog manually and leave DietPi-RAMlog active.

Journald (journalctl) can log persistent as well, easiest is to create the dir `/var/log/journal`, which leads to it's default auto mode logging there. If you disable DietPi-RAMlog then, it is persistent, with some default retention times AFAIK. See: https://manpages.debian.org/stretch/systemd/journald.conf.5.en.html
The problem is, that these journal files are not readable via text editor, but if systemctl is available there as well, it seems to be easy (I luckily never needed it): https://unix.stackexchange.com/a/200107

Btw: journald works as syslog daemon as well, thus using dockers `syslog` logging driver would have the same result. But the native `journald` driver should use some more direct/faster logging commands.

Still I am interested in the usual amount of logs, a container creates. Could you paste some (anonymized) entries from your container log or `journalctl -t docker` / `journalctl -u docker` with timestamps, so we get an idea?

**Comment by SaturnusDJ** [comment]
Yes I was sampling a log. 😃
[Docker_Home_Assistant_30_mins_log.txt](https://github.com/Fourdee/DietPi/files/2730561/Docker_Home_Assistant_30_mins_log.txt)

This is Home Assistant on default logging level.
For example why scan Bluetooth that much? Because if the car comes back the key to which the Bluetooth tracker is attached needs to be found quickly in order to immediately turn on the hallway light. Also note that Bluetooth LE devices for privacy deploy many MAC-addresses besides their normal one which potentially means 5 log lines per Bluetooth LE device each scan. Just an example.

I found that the Docker container log contains a lot more entries than the log that Home Assistant shows you from inside its webinterface. Which at first was a deception to me for the real log.

The worst is when there is a configuration error. It's possible that every x seconds a whole block of 20 lines of code will be logged. This happened for me in a lesser way with DSMR Reader. When the MQTT element is not configured, it will keep logging that it spawned a MQTT process but didn't find a configuration.

Meanwhile I've put the Home Assistant default logging level to warn and stopped the DSMR Reader MQTT process spawning. Except for on purpose Home Assistant debugging and testing, 4 days without logs for non Home Assistant containers.

It's absolutely clear that in no way DietPi or its config is to blame. It's just that software does not expect the storage to be a potential problem thus DietPi can be a life saver here.

**Comment by Fourdee** [comment]
@SaturnusDJ 

Many thanks for the report 👍 

@MichaIng 

> To sum up: I would go with your journald solution and add info to our online docs, how to (re)apply persistent disk logging, if desired.
@Fourdee
What do you think?

Yep, definitely 👍

**Comment by MichaIng** [comment]
Hmm, the amount of `INFO` logs IMO is too much for systemd. I mean its personal taste and how you use journalctl, but I personally regularly check it to get an overview about if everything is in order, any service, system or kernel errors and such, and this only works comfortable, if there are no continuous info logs, but at best warnings and system/kernel related entries.

But the logging mode can be adjusted:
- https://docs.docker.com/v17.09/engine/reference/commandline/dockerd/#daemon-configuration-file
- `"debug": false,` # If not yet the case
- `"log-level": "warn",`

Indeed there seems to be no way to customize the json-file log location. This will always be within the individual docker containers data dir. journald logging mode thus is the only RAMlog solution, AFAIK.

**Comment by MichaIng** [comment]
Log level applies to `dockerd` but is not passed through to `containerd`, thus all container logs revert to default level "info", which is IMO too much to flood journald with.

There was a patch to address exactly this issue, but it seems to have gone until current version: https://github.com/moby/moby/commit/aaa1392279ffe67a63581d2a43b3a3c7b0fc47b1
__________
containerd is a separate systemd unit thus has it's own config file:
Setting in `/etc/containerd/config.toml`
```
[debug]
  level = "warn"
```
works.

**Comment by MichaIng** [comment]
PR up: #3791
Works well on first test, good reason to finally add some online docs to inform users where to search for Docker and container logs:
- `journalctl -u docker`
- `journalctl -u containerd`
- or combined: `journalctl -u docker -u containerd`

Log level reduced to warnings to prevent journald flood.

Since the Docker config file does not exist by default, exposing it is a bonus by itself and we might want to add a link to docs inside so users easily find out how to further configure Docker or restore file logs 😉.

**EDIT: Online docs: https://dietpi.com/phpbb/viewtopic.php?p=9120#p9120**
- Firstrun/usage info might be great, but I don't use Docker actively so cannot say much 😄.
